File: 2020/day01.hs
import Data.Set (Set, empty, insert, member)
import qualified Data.Map.Strict as M (Map, (!), empty, insert, member)

main :: IO ()
main = do
  expenses <- map read . lines <$> readFile "input/day01"
  putStr "Part 1: "
  print $ findPairMultiplication empty expenses
  putStr "Part 2: "
  print $ findTripleMultiplication M.empty [] expenses

findPairMultiplication :: Set Int -> [Int] -> Int
findPairMultiplication prev (x : xs)
  | (2020 - x) `member` prev = x * (2020 - x)
  | otherwise = findPairMultiplication (insert x prev) xs

findTripleMultiplication :: M.Map Int Int -> [Int] -> [Int] -> Int
findTripleMultiplication sums prev (x : xs)
  | diff `M.member` sums = x * (sums M.! diff)
  | otherwise = findTripleMultiplication updatedSums (x : prev) xs
  where
    diff = 2020 - x
    updatedSums = updateSums prev sums
    updateSums (p : ps) = updateSums ps . M.insert (x + p) (x * p)
    updateSums _ = id

File: 2020/day01.ex
defmodule M do
  def find_pair_multiplication(prev, [value | values]) do
    complement = 2020 - value

    if MapSet.member?(prev, complement) do
      value * complement
    else
      find_pair_multiplication(MapSet.put(prev, value), values)
    end
  end

  def find_triple_multiplication(sums, prev, [value | values]) do
    complement = Map.get(sums, 2020 - value)

    if complement do
      value * complement
    else
      prev
      |> Enum.reduce(sums, &Map.put(&2, &1 + value, &1 * value))
      |> find_triple_multiplication([value | prev], values)
    end
  end
end

expenses =
  "input/day01"
  |> File.read!()
  |> String.trim()
  |> String.split("\n")
  |> Enum.map(&String.to_integer/1)

IO.inspect(M.find_pair_multiplication(MapSet.new(), expenses), label: "Part 1")
IO.inspect(M.find_triple_multiplication(%{}, [], expenses), label: "Part 2")
